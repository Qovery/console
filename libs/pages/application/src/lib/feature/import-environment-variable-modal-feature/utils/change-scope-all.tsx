import { APIVariableScopeEnum } from 'qovery-typescript-axios'
import { UseFormSetValue } from 'react-hook-form'

export function changeScopeForAll(
  value: APIVariableScopeEnum,
  setValue: UseFormSetValue<{ [key: string]: string }>,
  keys: (string | boolean)[]
): void {
  keys.forEach((key) => {
    setValue(key + '_scope', value?.toString())
  })
}
