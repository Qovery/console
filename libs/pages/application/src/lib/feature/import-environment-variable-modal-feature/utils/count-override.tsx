export function countOverride(data: { [key: string]: string }, existingEnvVarNames: (string | undefined)[]) {
  let count = 0
  Object.keys(data)
    .filter((key) => key.indexOf('_key') >= 0)
    .forEach((key) => {
      if (existingEnvVarNames.includes(data[key])) count++
    })

  return count
}
