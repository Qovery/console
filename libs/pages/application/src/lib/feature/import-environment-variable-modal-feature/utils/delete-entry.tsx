export function deleteEntry(
  key: string,
  setKeys: (keys: string[]) => void,
  keys: string[],
  unregister: (name: string | string[]) => void
): void {
  const newKeys = keys.filter((el) => el !== key)
  setKeys(newKeys)
  unregister(key + '_key')
  unregister(key + '_value')
  unregister(key + '_secret')
  unregister(key + '_scope')
}
