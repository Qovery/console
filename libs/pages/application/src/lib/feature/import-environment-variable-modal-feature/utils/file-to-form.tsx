import { APIVariableScopeEnum } from 'qovery-typescript-axios'

export function jsonToForm(json: string): { [key: string]: string } {
  const parsed = JSON.parse(json)
  return parsedToForm(parsed)
}

export function parsedToForm(parsed: { [key: string]: string }): { [key: string]: string } {
  const defaultValues: { [key: string]: string } = {}
  Object.keys(parsed).forEach((key) => {
    defaultValues[key + '_key'] = key || ''
    defaultValues[key + '_value'] = parsed[key] || ''
    defaultValues[key + '_scope'] = APIVariableScopeEnum.PROJECT
    defaultValues[key + '_secret'] = ''
  })

  return defaultValues
}
