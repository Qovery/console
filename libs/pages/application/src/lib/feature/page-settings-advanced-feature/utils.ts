import { ApplicationAdvancedSettings } from 'qovery-typescript-axios'
import { ApplicationEntity } from '@console/shared/interfaces'

export function initFormValues(keys: string[], application: ApplicationEntity): { [key: string]: string } {
  const values: { [key: string]: string } = {}
  keys.forEach((key) => {
    if (application.advanced_settings?.current_settings) {
      values[key] =
        application.advanced_settings.current_settings[key as keyof ApplicationAdvancedSettings]?.toString() || ''
    }
  })

  return values
}
