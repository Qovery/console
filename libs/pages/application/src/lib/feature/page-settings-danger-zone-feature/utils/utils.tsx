import { UseFormRegister, UseFormUnregister } from 'react-hook-form/dist/types/form'

export const onAddStorage = (
  register: UseFormRegister<{ [key: string]: string | number }>,
  keys: string[]
): string[] => {
  const id = Math.trunc(Math.random() * 10000).toString()

  register('size_' + id, { value: '' })
  register('type_' + id, { value: '' })
  register('path_' + id, { value: '' })

  return [...keys, id]
}

export const onRemove = (
  key: string,
  unregister: UseFormUnregister<{ [key: string]: string | number }>,
  keys: string[]
): string[] => {
  unregister('size_' + key)
  unregister('type_' + key)
  unregister('path_' + key)

  return keys.filter((k) => k !== key)
}
