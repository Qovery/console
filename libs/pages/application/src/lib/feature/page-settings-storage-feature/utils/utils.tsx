import { UseFormRegister, UseFormUnregister } from 'react-hook-form/dist/types/form'
import { ApplicationStorageStorage } from 'qovery-typescript-axios'

export const addStorage = (register: UseFormRegister<{ [key: string]: string | number }>, keys: string[]): string[] => {
  const id = Math.trunc(Math.random() * 10000).toString()

  register('size_' + id, { value: '' })
  register('type_' + id, { value: '' })
  register('path_' + id, { value: '' })

  return [...keys, id]
}

export const onRemove = (
  key: string,
  unregister: UseFormUnregister<{ [key: string]: string | number }>,
  keys: string[]
): string[] => {
  unregister('size_' + key)
  unregister('type_' + key)
  unregister('path_' + key)

  return keys.filter((k) => k !== key)
}

export const initStorage = (
  register: UseFormRegister<{ [key: string]: string | number }>,
  storages: ApplicationStorageStorage[]
): string[] => {
  const keys: string[] = []

  storages.forEach((storage) => {
    register('size_' + storage.id, { value: storage.size })
    register('type_' + storage.id, { value: storage.type })
    register('path_' + storage.id, { value: storage.mount_point })

    keys.push(storage.id || '')
  })

  return keys
}
