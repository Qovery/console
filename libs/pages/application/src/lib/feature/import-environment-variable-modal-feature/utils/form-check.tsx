import { APIVariableScopeEnum } from 'qovery-typescript-axios'
import { ServiceTypeEnum } from '@qovery/shared/enums'
import { EnvironmentVariableSecretOrPublic } from '@qovery/shared/interfaces'
import { getScopeHierarchy } from '@qovery/shared/utils'

export const validateKey = (
  value: string,
  existingVars: EnvironmentVariableSecretOrPublic[],
  currentScope: APIVariableScopeEnum,
  serviceType?: ServiceTypeEnum
): string | boolean => {
  if (value.toLowerCase().startsWith('qovery')) {
    return 'Variable name cannot begin with "QOVERY"'
  }

  const existingVar = existingVars.find((envVar) => envVar.key === value)
  if (existingVar && getScopeHierarchy(existingVar.scope, serviceType) > getScopeHierarchy(currentScope, serviceType)) {
    return 'This variable name already exists on a lower scope'
  }

  return true
}

export const warningMessage = (
  value: string,
  existingVars: EnvironmentVariableSecretOrPublic[],
  currentScope: APIVariableScopeEnum,
  overwriteEnabled = false,
  serviceType?: ServiceTypeEnum
): string | undefined => {
  const existingVar = existingVars.find((envVar) => envVar.key === value)

  if (existingVar && getScopeHierarchy(currentScope, serviceType) > getScopeHierarchy(existingVar.scope, serviceType)) {
    return 'This variable name already exists on a higher scope. An override will be created.'
  }

  if (existingVar && existingVar.scope === currentScope) {
    return overwriteEnabled
      ? 'This variable name on the same scope already exists and will be updated'
      : 'This variable already exists and will not be updated'
  }

  return undefined
}
