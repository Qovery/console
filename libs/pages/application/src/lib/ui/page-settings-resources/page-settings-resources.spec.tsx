import ResizeObserver from '__tests__/utils/resize-observer'
import { act, render, screen, waitFor } from '__tests__/utils/setup-jest'
import { wrapWithReactHookForm } from '__tests__/utils/wrap-with-react-hook-form'
import { applicationFactoryMock } from '@qovery/shared/factories'
import { ApplicationResourcesData } from '@qovery/shared/interfaces'
import { IconAwesomeEnum } from '@qovery/shared/ui'
import PageSettingsResources, { PageSettingsResourcesProps } from './page-settings-resources'

const application = applicationFactoryMock(1)[0]

const props: PageSettingsResourcesProps = {
  loading: false,
  onSubmit: () => jest.fn(),
  application: application,
  displayWarningCpu: true,
}

jest.mock('react-hook-form', () => ({
  ...jest.requireActual('react-hook-form'),
  useFormContext: () => ({
    watch: () => jest.fn(),
    formState: {
      isValid: true,
    },
  }),
}))

describe('PageSettingsResources', () => {
  window.ResizeObserver = ResizeObserver
  let defaultValues: ApplicationResourcesData

  beforeEach(() => {
    defaultValues = {
      instances: [1, 18],
      cpu: [3],
      memory: 1024,
    }
  })

  it('should render successfully', async () => {
    const { baseElement } = render(wrapWithReactHookForm(<PageSettingsResources {...props} />, { defaultValues }))
    expect(baseElement).toBeTruthy()
  })

  it('should render the form', async () => {
    const { getByDisplayValue } = render(
      wrapWithReactHookForm(<PageSettingsResources {...props} />, {
        defaultValues,
      })
    )

    const inputs = screen.getAllByRole('slider') as HTMLSpanElement[]

    await act(() => {
      getByDisplayValue(1024)
      expect(inputs[0].getAttribute('aria-valuenow')).toBe('3')
      expect(inputs[1].getAttribute('aria-valuenow')).toBe('1')
      expect(inputs[2].getAttribute('aria-valuenow')).toBe('18')
    })
  })

  it('should render warning box and icon for cpu', () => {
    props.displayWarningCpu = true

    const { getByTestId, getAllByRole } = render(
      wrapWithReactHookForm(<PageSettingsResources {...props} />, {
        defaultValues: { cpu: [10], instances: [1, 1], memory: 323 },
      })
    )

    const img = getAllByRole('img')[0]

    getByTestId('warning-box')
    expect(img.classList.contains(IconAwesomeEnum.TRIANGLE_EXCLAMATION)).toBe(true)
  })

  it('should submit the form', async () => {
    const spy = jest.fn((e) => e.preventDefault())
    props.onSubmit = spy
    props.loading = false

    render(
      wrapWithReactHookForm(<PageSettingsResources {...props} />, {
        defaultValues,
      })
    )

    const button = screen.getByTestId('submit-button')

    await waitFor(() => {
      button.click()
      expect(button).not.toBeDisabled()
      expect(spy).toHaveBeenCalled()
    })
  })
})
