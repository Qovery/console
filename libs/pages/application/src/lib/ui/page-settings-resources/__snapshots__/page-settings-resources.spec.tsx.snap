// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSettingsResources should render warning box and icon for cpu 1`] = `
<div>
  <div
    class="flex w-full flex-col justify-between"
  >
    <section
      class="flex flex-col max-w-content-with-navigation-left p-8"
    >
      <div
        class="mb-8 flex w-full justify-between gap-2"
      >
        <div>
          <h1
            class="text-neutral-400 text-xl font-bold mb-2"
          >
            Resources
          </h1>
          <p
            class="mb-3 max-w-lg text-sm text-neutral-400"
          >
            Manage the resources assigned to the service.
          </p>
          <button
            class="inline-flex shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-xs h-6 px-2 rounded bg-brand-50 border border-brand-500 text-brand-500 hover:[&:not(:active):not(:disabled)]:bg-white w-fit items-center gap-1"
          >
            <i
              aria-hidden="true"
              class="fa-regular fa-circle-question p-0.5 text-xs"
            />
            <span>
              Need help here?
            </span>
          </button>
        </div>
      </div>
      <form
        class="space-y-10"
      >
        <section
          class="flex flex-col gap-4"
        >
          <h2
            class="text-neutral-400 text-base font-bold"
          >
            Resources configuration
          </h2>
          <div
            class=""
            data-testid="input-text-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input    input--label-up "
              >
                <div
                  class=""
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="vCPU (milli)"
                  >
                    vCPU (milli)
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-text"
                    id="vCPU (milli)"
                    name="cpu"
                    type="number"
                    value="10"
                  />
                </div>
              </div>
            </div>
            <p
              class="mt-0.5 px-3 text-xs font-normal text-neutral-350 dark:text-neutral-50"
            >
              Minimum value is 
              10
               milli vCPU.
               
              Maximum value allowed based on the selected cluster instance type: 
              10
               milli vCPU.
               
            </p>
          </div>
          <div
            class="flex flex-row gap-x-3 p-3 border rounded text-sm border-red-500 bg-red-50 text-red-500 mt-3"
            data-accent-color="red"
            data-testid="banner-box"
          >
            <div
              class="-order-2 text-lg leading-6"
            >
              <i
                aria-hidden="true"
                class="fa-regular fa-triangle-exclamation "
              />
            </div>
            <div
              class="-order-1 mr-auto gap-x-3 text-neutral-400"
            >
              <span
                class="block font-medium leading-6"
              >
                Not enough resources
              </span>
              <span
                class="text-neutral-400"
              >
                Increase the capacity of your cluster nodes or reduce the service consumption.
              </span>
            </div>
          </div>
          <div
            class=""
            data-testid="input-memory-memory-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input    input--label-up "
              >
                <div
                  class=""
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="Memory (MiB)"
                  >
                    Memory (MiB)
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-memory-memory"
                    id="Memory (MiB)"
                    name="memory"
                    type="number"
                    value="323"
                  />
                </div>
              </div>
            </div>
            <p
              class="mt-0.5 px-3 text-xs font-normal text-neutral-350 dark:text-neutral-50"
            >
              Minimum value is 
              1
               MiB.
               
              Maximum value allowed based on the selected cluster instance type: 
              10
               MiB.
               
            </p>
          </div>
        </section>
        <section
          class="flex flex-col gap-4"
        >
          <h2
            class="text-neutral-400 text-base font-bold"
          >
            Instances
          </h2>
          <div
            class="grid grid-cols-2 gap-4"
          >
            <div
              class=""
              data-testid="input-text-wrapper"
            >
              <div
                class="relative"
              >
                <div
                  aria-label="input-container"
                  class="input    input--label-up "
                >
                  <div
                    class=""
                  >
                    <label
                      class="translate-y-2 text-sm"
                      for="Instances min"
                    >
                      Instances min
                    </label>
                    <input
                      class="input__value "
                      data-testid="input-text"
                      id="Instances min"
                      name="min_running_instances"
                      type="number"
                      value="1"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class=""
              data-testid="input-text-wrapper"
            >
              <div
                class="relative"
              >
                <div
                  aria-label="input-container"
                  class="input    input--label-up "
                >
                  <div
                    class=""
                  >
                    <label
                      class="translate-y-2 text-sm"
                      for="Instances max"
                    >
                      Instances max
                    </label>
                    <input
                      class="input__value "
                      data-testid="input-text"
                      id="Instances max"
                      name="max_running_instances"
                      type="number"
                      value="1"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <p
            class="text-xs text-neutral-350"
          >
            Application auto-scaling is based on real-time CPU consumption. When your app goes above 60% (default) of CPU consumption for 5 minutes, your app will be auto-scaled and more instances will be added.
          </p>
        </section>
        <div
          class="flex justify-end"
        >
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100"
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </section>
  </div>
</div>
`;
